File: sdk/java-sdk/src/test/java/com/evernym/verity/sdk/protocols/BasicMessageTest.java
package com.evernym.verity.sdk.protocols;

import com.evernym.verity.sdk.TestHelpers;
import com.evernym.verity.sdk.exceptions.VerityException;
import com.evernym.verity.sdk.protocols.basicmessage.BasicMessage;
import com.evernym.verity.sdk.protocols.basicmessage.v1_0.BasicMessageV1_0;
import com.evernym.verity.sdk.utils.Context;
import com.evernym.verity.sdk.utils.Util;
import org.json.JSONObject;
import org.junit.Test;

import static com.evernym.verity.sdk.TestHelpers.unpackForwardMessage;
import static org.junit.Assert.*;

public class BasicMessageTest {

    private final String forRelationship = "abcd12345";
    private final String content = "message text";
    private final String sent_time = "2018-12-13T17:29:34+0000";
    private final String localization = "en";

    @Test
    public void testGetMessageType() {
        BasicMessageV1_0 testProtocol = BasicMessage.v1_0(
                forRelationship,
                content,
                sent_time,
                localization);

        String msgName = "msg name";
        String msgType = Util.getMessageType(
                Util.COMMUNITY_MSG_QUALIFIER,
                testProtocol.family(),
                testProtocol.version(),
                msgName);
        assertEquals(msgType, testProtocol.messageType(msgName));
    }

    @Test
    public void testGetThreadId() {
        BasicMessageV1_0 testProtocol = BasicMessage.v1_0(
                forRelationship,
                content,
                sent_time,
                localization);
        assertNotNull(testProtocol.getThreadId());
    }

    @Test
    public void testConstructor() throws VerityException {
        Context context = TestHelpers.getContext();
        BasicMessageV1_0 testProtocol = BasicMessage.v1_0(
                forRelationship,
                content,
                sent_time,
                localization
        );
        JSONObject msg = testProtocol.messageMsg(context);
        testMessageMessages(msg);
    }

    private void testMessageMessages(JSONObject msg) {
        testBaseMessage(msg);
        assertEquals(forRelationship, msg.getString("~for_relationship"));
        assertEquals(content, msg.getString("content"));
        assertEquals(sent_time, msg.getString("sent_time"));
        assertEquals(localization, msg.getJSONObject("~l10n").getString("locale"));
    }

    @Test
    public void testMessage() throws Exception {
        Context context = null;
        try {
            context = TestHelpers.getContext();
            BasicMessageV1_0 testProtocol = BasicMessage.v1_0(
                    forRelationship,
                    content,
                    sent_time,
                    localization);

            byte[] message = testProtocol.messageMsgPacked(context);
            JSONObject unpackedMessage = unpackForwardMessage(context, message);
            testBaseMessage(unpackedMessage);
            assertEquals(
                    "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/basicmessage/1.0/send-message",
                    unpackedMessage.getString("@type")
            );
        } catch(Exception e) {
            e.printStackTrace();
            fail();
        } finally {
            TestHelpers.cleanup(context);
        }
    }

    private void testBaseMessage(JSONObject msg) {
        assertNotNull(msg.getString("@type"));
        assertNotNull(msg.getString("@id"));
        assertNotNull(msg.getJSONObject("~thread").getString("thid"));
    }
}

File: sdk/dotnet-sdk/src/VeritySDK.Test/Protocols/BasicMessage/BasicMessageTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Json;
using VeritySDK.Protocols.BasicMessage;
using VeritySDK.Utils;

namespace VeritySDK.Test
{
    [TestClass]
    public class BasicMessageTest : TestBase
    {
        private string forRelationship = "abcd12345";
        private string content = "Hello, World!";
        private string sentTime = "2018-12-13T17:29:34+0000";
        private string localization = "en";

        [TestMethod]
        public void testGetMessageType()
        {
            BasicMessageV1_0 testProtocol = BasicMessage.v1_0(
                    forRelationship,
                    content,
                    sentTime,
                    localization);

            string msgName = "msg name";
            string msgType = Util.getMessageType(
                    Util.COMMUNITY_MSG_QUALIFIER,
                    testProtocol.family(),
                    testProtocol.version(),
                    msgName);
            Assert.AreEqual(msgType, testProtocol.messageType(msgName));
        }

        [TestMethod]
        public void testGetThreadId()
        {
            BasicMessageV1_0 testProtocol = BasicMessage.v1_0(
                    forRelationship,
                    content,
                    sentTime,
                    localization);
            Assert.IsNotNull(testProtocol.getThreadId());
        }

        [TestMethod]
        public void testConstructor()
        {
            Context context = TestHelpers.getContext();
            BasicMessageV1_0 testProtocol = BasicMessage.v1_0(
                    forRelationship,
                    content,
                    sentTime,
                    localization
            );
            JsonObject msg = testProtocol.messageMsg(context);
            testSendMessages(msg);
        }

        private void testSendMessages(JsonObject msg)
        {
            testBaseMessage(msg);
            Assert.AreEqual(forRelationship, msg.getAsString("~for_relationship"));
            Assert.AreEqual(content, msg.getAsString("content"));
            Assert.AreEqual(sentTime, msg.getAsString("sent_time"));
            Assert.AreEqual(localization, msg.getAsJsonObject("~l10n").getAsString("locale"));
        }

        [TestMethod]
        public void testSend()
        {
            withContext(context => {
                BasicMessageV1_0 testProtocol = BasicMessage.v1_0(
                    forRelationship,
                    content,
                    sentTime,
                    localization);

                byte[] message = testProtocol.messageMsgPacked(context);
                JsonObject unpackedMessage = TestHelpers.unpackForwardMessage(context, message);
                testBaseMessage(unpackedMessage);
                Assert.AreEqual(
                        "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/basicmessage/1.0/send-message",
                        unpackedMessage.getAsString("@type")
                );
            });
        }

        private void testBaseMessage(JsonObject msg)
        {
            Assert.IsNotNull(msg.getAsString("@type"));
            Assert.IsNotNull(msg.getAsString("@id"));
            Assert.IsNotNull(msg.getAsJsonObject("~thread").getAsString("thid"));
        }
    }
}
